sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala

```diff
@@ -502,21 +500,20 @@ case class WholeStageCodegenRDD(@transient sc: SparkContext, var source: CodeAnd
 
   override def compute(split: Partition,
       context: TaskContext): Iterator[InternalRow] = {
-    val catcher = catching(classOf[ClassCastException])
     new Iterator[InternalRow] {
-      private[this] var i = computeInternal(split, context)
-
-      private[this] def replace() = i = {
-        logInfo(s"ClassCast Exception, hence recompiling")
-        CodeGenerator.invalidate(source)
-        computeInternal(split, context)
-      }
-
-      override def hasNext: Boolean = catcher.opt(i.hasNext).getOrElse {
-        replace(); hasNext
+      private[this] var iter = computeInternal(split, context)
+
+      override def hasNext: Boolean = try {
+        iter.hasNext
+      } catch {
+        case _: ClassCastException =>
+          logInfo(s"ClassCastException, hence recompiling")
+          CodeGenerator.invalidate(source)
+          iter = computeInternal(split, context)
```

Comment:
I read in one of @sumwale comment that we should try only once as there could be genuine CCE. Is this right? This change looks like will recompile as long as hasNext returns truue. Isn't it. Please explain. May be I am missing something.